src/systems/particle_emitter/ParticleEmitter.cc

```diff
@@ -0,0 +1,186 @@
+{
+  /// \brief The particle emitter.
+  public: ignition::msgs::ParticleEmitter emitter;
+};
+
+//////////////////////////////////////////////////
+ParticleEmitter::ParticleEmitter()
+  : System(), dataPtr(std::make_unique<ParticleEmitterPrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void ParticleEmitter::Configure(const Entity &/*_entity*/,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    EntityComponentManager &_ecm,
+    EventManager &/*_eventMgr*/)
+{
+  // Create a particle emitter entity.
+  auto entity = _ecm.CreateEntity();
+  if (entity == kNullEntity)
+  {
+    ignerr << "Failed to create a particle emitter entity" << std::endl;
+    return;
+  }
+
+  // Name.
+  std::string name = "particle_emitter_entity_" + std::to_string(entity);
+  if (_sdf->HasElement("emitter_name"))
+    name = _sdf->Get<std::string>("emitter_name");
+  this->dataPtr->emitter.set_name(name);
+
+  // Type. The default type is point.
+  this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_POINT);
+  std::string type = _sdf->Get<std::string>("type", "point").first;
+  if (type == "box")
+  {
+    this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_BOX);
+  }
+  else if (type == "cylinder")
+  {
+    this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_CYLINDER);
+  }
+  else if (type == "ellipsoid")
+  {
+    this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_ELLIPSOID);
+  }
+  else if (type != "point")
+  {
+    ignerr << "Unknown emitter type [" << type << "]. Using [point] instead"
+           << std::endl;
+  }
+
+  // Pose.
+  ignition::math::Pose3d pose =
+    _sdf->Get<ignition::math::Pose3d>("pose");
+  ignition::msgs::Set(this->dataPtr->emitter.mutable_pose(), pose);
+
+  // Size.
+  ignition::math::Vector3d size = ignition::math::Vector3d::One;
+  if (_sdf->HasElement("size"))
+    size = _sdf->Get<ignition::math::Vector3d>("size");
+  ignition::msgs::Set(this->dataPtr->emitter.mutable_size(), size);
+
+  // Rate.
+  this->dataPtr->emitter.set_rate(_sdf->Get<double>("rate", 10).first);
+
+  // Duration.
+  this->dataPtr->emitter.set_duration(_sdf->Get<double>("duration", 0).first);
+
+  // Emitting.
+  this->dataPtr->emitter.set_emitting(_sdf->Get<bool>("emitting", false).first);
+
+  // Particle size.
+  size = ignition::math::Vector3d::One;
+  if (_sdf->HasElement("particle_size"))
+    size = _sdf->Get<ignition::math::Vector3d>("particle_size");
+  ignition::msgs::Set(this->dataPtr->emitter.mutable_particle_size(), size);
+
+  // Lifetime.
+  this->dataPtr->emitter.set_lifetime(_sdf->Get<double>("lifetime", 5).first);
+
+  // Material.
+  if (_sdf->HasElement("material"))
+  {
+    auto materialElem = _sdf->GetElementImpl("material");
+    sdf::Material material;
+    material.Load(materialElem);
+    ignition::msgs::Material materialMsg = convert<msgs::Material>(material);
```

Comment:
It looks to me like this is never used anywhere else, does `materialMsg` need to be passed along somewhere/somehow in order to use the specified material?

Reply:
Edit: it looks like the material is being set in #593: https://github.com/ignitionrobotics/ign-gazebo/blob/502abfa1a1d2515a71a66bffe86bdae2a06c7e72/src/systems/particle_emitter/ParticleEmitter.cc#L134